File: packages/djd-agent-score-mcp/src/tools.ts
/**
 * MCP tool registrations for the DJD Agent Score API.
 *
 * Each tool maps to one API endpoint. Free tools work without an API key
 * (10 queries/day); paid tools require DJD_API_KEY with credits.
 */

import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js'
import { z } from 'zod'

import type { DJDClient } from './client.js'
import type {
  BasicScoreResponse,
  FullScoreResponse,
  LeaderboardResponse,
  EconomyMetrics,
} from 'djd-agent-score-client'
import {
  formatBasicScore,
  formatFullScore,
  formatHistory,
  formatLeaderboard,
  formatEconomyMetrics,
  formatBatchScore,
} from './format.js'
import type { HistoryResponse } from './format.js'

// ── Helpers ──────────────────────────────────────────────────────────────────

function textResult(text: string) {
  return { content: [{ type: 'text' as const, text }] }
}

function errorResult(message: string) {
  return { content: [{ type: 'text' as const, text: `Error: ${message}` }], isError: true as const }
}

async function callTool<T>(fn: () => Promise<T>, formatter: (data: T) => string) {
  try {
    const data = await fn()
    return textResult(formatter(data))
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err)
    return errorResult(message)
  }
}

// ── Wallet schema (reused across tools) ──────────────────────────────────────

const walletSchema = z
  .string()
  .regex(/^0x[a-fA-F0-9]{40}$/, 'Must be a valid Ethereum address (0x + 40 hex chars)')

// ── Tool registration ────────────────────────────────────────────────────────

export function registerTools(server: McpServer, client: DJDClient): void {
  // 1. Basic score (free, 10/day)
  server.tool(
    'get_score',
    'Get the DJD reputation score for an Ethereum wallet. Returns score (0-100), trust tier, confidence level, and recommendation. Free tier: 10 queries/day.',
    { wallet: walletSchema },
    async ({ wallet }) =>
      callTool(
        () => client.get<BasicScoreResponse>(`/v1/score/basic?wallet=${wallet}`),
        formatBasicScore,
      ),
  )

  // 2. Full score with dimensions ($0.10 — needs API key)
  server.tool(
    'get_full_score',
    'Get detailed DJD reputation breakdown: reliability, viability, identity, capability dimensions plus sybil detection, gaming indicators, and improvement paths. Requires API key ($0.10/query).',
    { wallet: walletSchema },
    async ({ wallet }) =>
      callTool(
        () => client.get<FullScoreResponse>(`/v1/score/full?wallet=${wallet}`),
        formatFullScore,
      ),
  )

  // 3. Score history with trajectory ($0.15 — needs API key)
  server.tool(
    'get_score_history',
    'Get historical score data for a wallet with trend analysis and trajectory metrics (velocity, momentum, direction). Requires API key ($0.15/query).',
    {
      wallet: walletSchema,
      limit: z.number().int().min(1).max(100).default(50).describe('Max entries to return (1-100)'),
      after: z.string().optional().describe('Only scores after this ISO date (YYYY-MM-DD)'),
      before: z.string().optional().describe('Only scores before this ISO date (YYYY-MM-DD)'),
    },
    async ({ wallet, limit, after, before }) => {
      let path = `/v1/score/history?wallet=${wallet}&limit=${limit}`
      if (after) path += `&after=${after}`
      if (before) path += `&before=${before}`

      return callTool(
        () => client.get<HistoryResponse>(path),
        formatHistory,
      )
    },
  )

  // 4. Leaderboard (free)
  server.tool(
    'get_leaderboard',
    'Get the DJD agent leaderboard — top-scoring wallets ranked by reputation. Includes registration status and GitHub verification. Free endpoint.',
    {},
    async () =>
      callTool(
        () => client.get<LeaderboardResponse>('/v1/leaderboard'),
        formatLeaderboard,
      ),
  )

  // 5. Economy metrics (free)
  server.tool(
    'get_economy_metrics',
    'Get DJD network health metrics: scoring volume, agent registrations, and economy stats. Free endpoint.',
    {
      period: z.enum(['daily', 'weekly', 'monthly']).default('daily').describe('Aggregation period'),
      limit: z.number().int().min(1).max(90).default(30).describe('Number of periods to return'),
    },
    async ({ period, limit }) =>
      callTool(
        () => client.get<EconomyMetrics>(`/v1/data/economy?period=${period}&limit=${limit}`),
        formatEconomyMetrics,
      ),
  )

  // 6. Batch score (2-20 wallets, $0.50 — needs API key)
  server.tool(
    'batch_score',
    'Score multiple wallets in one request (2-20 addresses). Returns basic score for each. Requires API key ($0.50/batch).',
    {
      wallets: z
        .array(walletSchema)
        .min(2)
        .max(20)
        .describe('Array of 2-20 Ethereum wallet addresses to score'),
    },
    async ({ wallets }) =>
      callTool(
        () => client.post<{ results: BasicScoreResponse[]; count: number }>('/v1/score/batch', { wallets }),
        formatBatchScore,
      ),
  )
}

File: integrations/crewai/tools.py
"""CrewAI tools for querying the DJD AgentScore API.

Each tool is a standalone function decorated with ``@tool``, following
CrewAI's idiomatic pattern.  All tools are synchronous (CrewAI runs
tools in a sync context) but use ``httpx`` for clean HTTP handling.

Usage in a CrewAI agent::

    from crewai import Agent
    from integrations.crewai.tools import check_agent_score, get_detailed_agent_score

    agent = Agent(
        role="Trust Analyst",
        goal="Verify wallet reputation before transactions",
        tools=[check_agent_score, get_detailed_agent_score],
    )
"""

from __future__ import annotations

import json
import os
from typing import Optional

import httpx
from crewai.tools import tool

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

BASE_URL = os.getenv("DJD_BASE_URL", "https://djdagentscore.dev")
API_KEY = os.getenv("DJD_API_KEY", "")
TIMEOUT = 30.0


def _get_headers() -> dict[str, str]:
    """Build request headers, including API key if configured."""
    headers = {"Accept": "application/json"}
    key = API_KEY or os.getenv("DJD_API_KEY", "")
    if key:
        headers["Authorization"] = f"Bearer {key}"
    return headers


def _api_get(path: str, params: Optional[dict[str, str]] = None) -> dict:
    """Synchronous GET request to the AgentScore API."""
    resp = httpx.get(
        f"{BASE_URL}{path}",
        params=params,
        headers=_get_headers(),
        timeout=TIMEOUT,
    )
    resp.raise_for_status()
    return resp.json()


def _api_post(path: str, body: dict) -> dict:
    """Synchronous POST request to the AgentScore API."""
    resp = httpx.post(
        f"{BASE_URL}{path}",
        json=body,
        headers=_get_headers(),
        timeout=TIMEOUT,
    )
    resp.raise_for_status()
    return resp.json()


# ---------------------------------------------------------------------------
# Tool: Basic Score
# ---------------------------------------------------------------------------

@tool("Check Agent Reputation Score")
def check_agent_score(wallet: str) -> str:
    """Look up the DJD AgentScore reputation rating for an Ethereum wallet address.

    Returns a trust score (0-100), tier (Elite/Trusted/Established/Emerging/Unverified),
    confidence level, and a human-readable recommendation.

    This is the FREE endpoint — no API key required. Use this to quickly check
    if an AI agent wallet is trustworthy before transacting with it.

    Args:
        wallet: Ethereum wallet address (0x-prefixed, 40 hex chars).
    """
    try:
        data = _api_get("/v1/score/basic", params={"wallet": wallet})
        return json.dumps(data, indent=2)
    except httpx.HTTPStatusError as e:
        return f"API error {e.response.status_code}: {e.response.text}"
    except Exception as e:
        return f"Request failed: {e}"


# ---------------------------------------------------------------------------
# Tool: Detailed Score
# ---------------------------------------------------------------------------

@tool("Get Detailed Agent Score")
def get_detailed_agent_score(wallet: str) -> str:
    """Get a detailed reputation breakdown for an AI agent wallet on Base L2.

    Returns the full scoring dimensions (reliability, viability, identity,
    capability, behavior), sybil detection flags, gaming indicators,
    score history, trajectory analysis, and improvement suggestions.

    IMPORTANT: This is a PAID endpoint ($0.10/query). Set the DJD_API_KEY
    environment variable before using this tool.

    Args:
        wallet: Ethereum wallet address (0x-prefixed, 40 hex chars).
    """
    try:
        data = _api_get("/v1/score/full", params={"wallet": wallet})
        return json.dumps(data, indent=2)
    except httpx.HTTPStatusError as e:
        return f"API error {e.response.status_code}: {e.response.text}"
    except Exception as e:
        return f"Request failed: {e}"


# ---------------------------------------------------------------------------
# Tool: Agent Registration
# ---------------------------------------------------------------------------

@tool("Register AI Agent")
def register_agent(
    wallet: str,
    name: str = "",
    description: str = "",
    github_url: str = "",
    website_url: str = "",
) -> str:
    """Register an AI agent wallet with the DJD AgentScore network.

    Registration is FREE and optional, but registered agents appear on the
    public leaderboard and can earn a GitHub-verified badge. Providing a
    public GitHub repo URL triggers automatic verification.

    Args:
        wallet: Ethereum wallet address to register.
        name: Optional display name for the agent (max 100 chars).
        description: Optional description of the agent's purpose (max 500 chars).
        github_url: Optional HTTPS GitHub repo URL for source code.
        website_url: Optional HTTPS URL for the agent's website.
    """
    body: dict = {"wallet": wallet}
    if name:
        body["name"] = name
    if description:
        body["description"] = description
    if github_url:
        body["github_url"] = github_url
    if website_url:
        body["website_url"] = website_url

    try:
        data = _api_post("/v1/agent/register", body)
        return json.dumps(data, indent=2)
    except httpx.HTTPStatusError as e:
        return f"API error {e.response.status_code}: {e.response.text}"
    except Exception as e:
        return f"Request failed: {e}"


# ---------------------------------------------------------------------------
# Tool: Leaderboard
# ---------------------------------------------------------------------------

@tool("Get Agent Leaderboard")
def get_agent_leaderboard() -> str:
    """Retrieve the DJD AgentScore public leaderboard.

    Returns the top-ranked AI agent wallets with their scores, tiers,
    registration status, and GitHub verification badges. This is the
    FREE endpoint — no API key required.

    Use this to find the most trusted agents on Base L2 or to see where
    a specific agent ranks relative to others.
    """
    try:
        data = _api_get("/v1/leaderboard")
        return json.dumps(data, indent=2)
    except httpx.HTTPStatusError as e:
        return f"API error {e.response.status_code}: {e.response.text}"
    except Exception as e:
        return f"Request failed: {e}"
